specs/data-store/deletion-and-deprecation.md

```diff
@@ -0,0 +1,231 @@
+   physical blob deletions per invocation. *N* should be configurable with a
+   default of 10.
+
+   The deletion daemon can optionally be invoked in "dry run" mode in which it
+   simply logs the actions it would perform in a wet run. In dry-run mode, *N*
+   is ignored.
+
+   The deletion daemon is idempotent. Already deleted blobs and already
+   existing deletion markers do not count against *N*.
+
+### Primary indexes
+
+Bundle deletion markers are indexed verbatim by the Data Store. The JSON schema
+for deletion markers is defined such that there are no conflicts with the
+top-level keys in the index documents for regular bundles (`files`, `manifest`,
+`uuid`, `version` etc). Bundle deletion markers are therefore compatible with
+regular bundle documents with respect to in the same index even if
+Elasticsearch's dynamic mapping is enabled.
+
+### 1.3. Replication
+
+The replication process in Data Store replicates file and bundle deletion
+markers. It does not replicate physical deletion of blobs. To physically delete
+blobs in all replicas, an administrator needs to repeatedly run the Data Store
+admin CLI to initiate a deletion process against each of the replicas.
+
+### 1.4. Consistency
+
+It is the administrator's responsibility to enumerate all bundles referring to
+a file that is to be deleted and issue `DELETE /bundles` requests for each of
+them. If the administrator misses such a bundle (a *dangling bundle*), users
+requesting `GET /bundle` for a dangling bundle will be able to retrieve its
+manifest. Likewise, `POST /search` requests will return the metadata for
+dangling bundles. Secondary indexes like the Data Browser (Orange Box) will
+continue to hand out (meta)data for dangling bundles. If users request `GET
+/files` for a file contained in a dangling bundle and the requested file was
+referenced by another already logically (physically) deleted bundle, the Data
+Store will respond with a 404 (500) status code.
+
+The administrator can initiate a consistency check against a storage replica to
+list dangling bundles. The administrator can then issue more `DELETE /bundles`
+requests for those dangling bundles.
+
+### Secondary indexes
+
+The Data Store can optionally notify subscribers about the creation of a bundle
+deletion marker. Secondary index services need to register a subscription using
+a query that matches deletion markers. When they are notified about a deletion
+marker, they need to take the necessary actions to remove the data from their
+database or index.
+
+### Derived data
+
+The metadata in bundles containing derived data must reference the input bundle
+versions. Administrators are responsible for tracking down data bundles with
+data derived from data in deleted bundles and to issue deletion requests for
+them.
+
+### Tiered storage
+
+Physical deletion of blobs is effective across all tiered storage classes in
+the underlying blob store (S3, GCP, ...).
+
+TBD: Does deletion from Glacier involve a delay?
```

Comment:
It does not, but we don't currently plan to use Glacier. (We plan to use S3-IA only for cost control on S3)